operators/src/pro/executor/replay.rs

```diff
@@ -0,0 +1,736 @@
+        self.senders
+            .get_mut(&sender_id)
+            .unwrap()
+            .update_waker(waker);
+    }
+
+    fn remove_send_waker(&mut self, sender_id: u32) {
+        self.senders.get_mut(&sender_id).unwrap().remove_waker();
+    }
+
+    fn create_receiver(&mut self, first: bool) -> Result<u32, SubscribeError> {
+        if self.offset > 0 {
+            return Err(SubscribeError::WouldLag);
+        } else if !first && self.receivers.is_empty() {
+            return Err(SubscribeError::Closed);
+        }
+
+        // Increment expected read count
+        for v in &mut self.buffer {
+            v.expected_receivers += 1;
+        }
+
+        let id = self.consumer_id_seq;
+        self.consumer_id_seq += 1;
+
+        assert!(self
+            .receivers
+            .insert(
+                id,
+                InnerReceiver {
+                    idx: 0,
+                    waker: None,
+                },
+            )
+            .is_none());
+        Ok(id)
+    }
+
+    fn remove_receiver(&mut self, id: u32) {
+        if let Some(r) = self.receivers.remove(&id) {
+            let idx = r.idx - self.offset;
+            let mut notify = false;
+            for i in idx..self.buffer.len() {
+                let e = &mut self.buffer[i];
+                e.expected_receivers -= 1;
+                notify |= e.expected_receivers == 0;
+            }
+            if notify {
+                self.notify_senders();
+            }
+        }
+    }
+
+    fn try_recv(&mut self, receiver_id: u32) -> Result<T, ReceiveError> {
+        let c = self.receivers.get_mut(&receiver_id).unwrap();
+        let q_idx = c.idx - self.offset;
+
+        let res = match self.buffer.get_mut(q_idx) {
+            Some(e) => {
+                e.expected_receivers -= 1;
+                Ok(e.value.clone())
+            }
+            None if self.senders.is_empty() => Err(ReceiveError::Closed),
+            None => Err(ReceiveError::NoMessage),
+        };
+
+        match res {
+            Ok(v) => {
+                c.idx += 1;
+                self.notify_senders();
+                Ok(v)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn notify_receivers(&mut self) {
+        for v in self.receivers.values_mut() {
+            if let Some(w) = v.waker.take() {
+                w.wake();
+            }
+        }
+    }
+
+    fn remove_receiver_waker(&mut self, receiver_id: u32) {
+        self.receivers.get_mut(&receiver_id).unwrap().remove_waker();
+    }
+
+    fn update_receiver_waker(&mut self, receiver_id: u32, waker: &Waker) {
+        self.receivers
+            .get_mut(&receiver_id)
+            .unwrap()
+            .update_waker(waker);
+    }
+}
+
+// /////////////////////////////////////////////////////////////////////////////
+//
+// SENDER SIDE
+//
+// /////////////////////////////////////////////////////////////////////////////
```

Comment:
Usually, doing this indicates the need of a second file :smile:

Reply:
Yes, but they belong closely together and it is not that much code. However, I can split it up if you want to.